### DISCUSSION TITLE
How do Targets Work?

### DISCUSSION THREAD

**Discussion by Squidcod** [cloudprober/cloudprober#315] [Q&A] [answered]
I'm trying to run the same external probe against multiple different domains. eg google.com yahoo.com msn.com

The documentation is relatively sparse with examples on how to use targets, but it seems like it's the correct place to do the thing.

- Does the following look correct to achieve this goal?
- Will it act as if there were 3 separate probe entries but without having to copy paste? Or will it have an effect on the way the metrics work?
- Can I use the additional label field as I do below?

EDIT: Updated the original question to have a working probe for easier viewing.

probe {
name: "test"
type: EXTERNAL
targets {
host_names: "hello.com, world.com, foo.com"
}
interval_msec: 60000
timeout_msec: 15000
external_probe {
mode: ONCE
command: "python3 /test.py --target=@host_names@"
}
additional_label {
key: "check_type"
value: "CI"
}
}

**Comment by manugarg** [comment, marked as answer]
Hello @Squidcod,

Sorry for the sparse documentation regarding how targets work.

* This will run 3 (sub)probes in parallel, each probe using a different target. 
* All probe metrics have a "dst" label (though we should've probably named it "target", but Cloudprober's earliest use case was network based probing, though it has come a long way since then):
```
   success{ptype="external", probe="test_install", dst="google.com"}
   total{...}
   latency{...}
   success{ptype="external", probe="test_install", dst="yahoo.com"}
   total{...}
   latency{...}
```
* Using additional label will add another label to all the metrics. So in the example above, you'll have additional label "target" that will have same values as "dst".

Update, Apr 20, 2023:
I've added some documentation to explain how targets work:
https://cloudprober.org/docs/how-to/targets/

**Reply by Squidcod** [reply]
Thanks for this answer.

**Reply by Squidcod** [reply]
As a heads up on trying I did get an error with my example, I'm starting investigation on fixing it, so do not use my code as an example.
`5641 cloudprober.go:182] Error initializing cloudprober. Err: line 4: unknown field name "target" in proto.TargetsDef`

**Reply by manugarg** [reply]
Apologies for missing it earlier, targets look like this:
targets {
  host_names: "google.com","yahoo.com", "msn.com"
}

Ex: https://github.com/cloudprober/cloudprober/blob/e441335ed11f4aec93becee7d8d290b1d80ef781/examples/surfacers/default_surfacers.cfg#L26

**Reply by Squidcod** [reply]
Thanks for the quick reply.
I'll give that a go.

I was basing it off of the Note in https://cloudprober.org/how-to/external-probe/#configuration
Which suggested supported fields.

**Reply by Squidcod** [reply]
Hrm

        type: EXTERNAL
        targets { 
            host_names: "hello.com","world.com" 
        }

results in 
cloudprober.go:182] Error initializing cloudprober. Err: line 5: unknown field name "\"world.com\"" in proto.TargetsDef

**Reply by manugarg** [reply]
It has to be one string: "google.com, yahoo.com, msn.com"

(Sorry, my bad again)

**Reply by Squidcod** [reply]
Awesome thank you! I see metrics coming in now.
Just to resolve the whole thing I'll put in what I've got (with some internal details altered)

probe {
        name: "test"
        type: EXTERNAL
        targets { 
            host_names: "hello.com, world.com" 
        }
        interval_msec: 60000
        timeout_msec: 15000
        external_probe {
                mode: ONCE
                command: "python3 /test.py --target=@host_names@"
        }
        additional_label {
                key: "check_type"
                value: "CI"
        }
}
